tutorials/13-seasonal-time-series/13_seasonal_time_series.jmd

```diff
@@ -0,0 +1,308 @@
+tmax = 10
+tt = 0:0.05:tmax
+f₁(t) = -1 + 2 * t
+f₂(t) = 5 * sin(2π * t * true_sin_freq / tmax)
+f₃(t) = 2.5 * cos(2π * t * true_cos_freq / tmax)
+f(t) = f₁(t) + f₂(t) + f₃(t)
+lw = 1
+
+plt_target = plot(
+    f, tt; label="f(t)", title="Observed time series", legend=:topleft, color=1, linewidth=3
+)
+plot!(plt_target, f₁, tt; label="f₁(t)", color=2, style=:dot, linewidth=lw)
+plot!(plt_target, f₂, tt; label="f₂(t)", color=3, style=:dash, linewidth=lw)
+plot!(plt_target, f₃, tt; label="f₃(t)", color=4, style=:dashdot, linewidth=lw)
+```
+
+Even though we use simple components, combining them can give rise to fairly complex time series.
+In this time series, cyclic effects are just added on top of the trend.
+If we instead multiply the components we get much more varying time series since all effect amplify each other.
+Note that we add 1 to the terms to keep the oscillations around the trend.
+
+```julia
+g(t) = f₁(t) * (1 + f₂(t)) * (1 + f₃(t))
+
+plt_target = plot(
+    g, tt; label="f(t)", title="Observed time series", legend=:topleft, color=1, linewidth=3
+)
+plot!(plt_target, f₁, tt; label="f₁(t)", color=2, style=:dot, linewidth=lw)
+plot!(plt_target, f₂, tt; label="f₂(t)", color=3, style=:dash, linewidth=lw)
+plot!(plt_target, f₃, tt; label="f₃(t)", color=4, style=:dashdot, linewidth=lw)
+```
+
+# Model fitting
+
+Having discussed time series decomposition, let us fit a model to the time series above and recover the true parameters.
+Before building our model, we standardise the time axis to $[0, 1]$ and subtract the max of the time series.
+This helps convergence while maintaining interpretability and the correct scales for the cyclic components.
+
+```julia
+σ = 0.35
+t = collect(tt)[begin:3:end]
+t_min, t_max = extrema(t)
+x = (t .- t_min) ./ (t_max - t_min)
+yf = f.(t) .+ σ * randn(size(t))
+yf_max = maximum(yf)
+yf = yf .- yf_max
+
+scatter(x, yf; title="Standardised data", label=nothing)
+```
+
+Let us now build our model.
+We want to assume a linear trend, and cyclic effects.
+Encoding a linear trend is easy enough, but what about cyclical effects?
+We will take a scattergun approach, and create multiple cyclical features using both sine and cosine functions and let our inference machinery figure out which to keep.
+To do this, we define how long a one period should be, and create features in reference to said period.
+How long a period should be is problem dependent, but as an example let us say it is $1$ year.
+If we then find evidence for a cyclic effect with a frequency of 2, that would mean a biannual effect. A frequency of 4 would mean quarterly etc.
+Since we are using synthetic data, we are simply going to let the period be 1, which is the entire length of the time series.
+
+```julia
+freqs = 1:10
+num_freqs = length(freqs)
+period = 1
+cyclic_feature(x, freqs, p, fn) = mapreduce(f -> fn.(2 * π * f * x / p), hcat, freqs)
+c = hcat(cyclic_feature(x, freqs, period, sin), cyclic_feature(x, freqs, period, cos))
```

Comment:
A bit simpler:
```suggestion
cyclic_features = [sinpi.(2 .* freqs' .* x) cospi.(2 .* freqs' .* x)]
```

Reply:
I left the period variable in the code to make it easier to adapt for new users who might want a different period